Mask refactor

shaungrady opened the pull request:
No more dependencies. Not even jQuery.
I've never written an end-to-end test before, so I'll need assistance rewriting the spec.

Big things to note about the refactor:
- $observe is set on the mask attr, so mask is fully dynamic.
- Model is now sent the unmasked value of the input. This is more useful, arguably, than the masked value for insertion into a database. Masking is a function of the view, not the model.

This is my first real directive I've written, so I'd love all the feedback I can get.
Thanks.

shaungrady committed:
Mask refactor, no more dependencies

shaungrady committed:
No more dependencies

petebacondarwin committed:
fix(mask): Get unit tests working, add demo page, allow interpolated masks

petebacondarwin commented:
Hi Shaun
This is excellent stuff!
I pulled your branch and made some changes:
https://github.com/petebacondarwin/angular-ui/tree/mask
See what you think.  Still not sure how to test UI triggered changes.
Pete

On 10 March 2013 06:32, Shaun Grady notifications@github.com wrote:

> No more dependencies. Not even jQuery.
> I've never written an end-to-end test before, so I'll need assistance
> rewriting the spec.
> 
> Big things to note about the refactor:
> - $observe is set on the mask attr, so mask is fully dynamic.
> - Model is now sent the unmasked value of the input. This is more
>   useful, arguably, than the masked value for insertion into a database.
>   Masking is a function of the display, not storage.
> 
> This is my first real directive I've written, so I'd love all the feedback
> I can get.
> 
> ## Thanks.
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/shaungrady/angular-ui mask
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/angular-ui/angular-ui/pull/479
> Commit Summary
> - Mask refactor, no more dependencies
> - No more dependencies
> 
> File Changes
> - _D_ modules/directives/mask/dependencies.jsonhttps://github.com/angular-ui/angular-ui/pull/479/files#diff-0(5)
> - _M_ modules/directives/mask/mask.jshttps://github.com/angular-ui/angular-ui/pull/479/files#diff-1(341)
> 
> Patch Links:
> - https://github.com/angular-ui/angular-ui/pull/479.patch
> - https://github.com/angular-ui/angular-ui/pull/479.diff

petebacondarwin commented:
By the way, I think that dynamic masks should be done with interpolation,
which removes the need to have single quote around every static mask:

ui-mask="{{someMaskVar}}"

On 11 March 2013 13:04, Peter Bacon Darwin pete@bacondarwin.com wrote:

> Hi Shaun
> This is excellent stuff!
> I pulled your branch and made some changes:
> https://github.com/petebacondarwin/angular-ui/tree/mask
> See what you think.  Still not sure how to test UI triggered changes.
> Pete
> 
> On 10 March 2013 06:32, Shaun Grady notifications@github.com wrote:
> 
> > No more dependencies. Not even jQuery.
> > I've never written an end-to-end test before, so I'll need assistance
> > rewriting the spec.
> > 
> > Big things to note about the refactor:
> > - $observe is set on the mask attr, so mask is fully dynamic.
> > - Model is now sent the unmasked value of the input. This is more
> >   useful, arguably, than the masked value for insertion into a database.
> >   Masking is a function of the display, not storage.
> > 
> > This is my first real directive I've written, so I'd love all the
> > feedback I can get.
> > 
> > ## Thanks.
> > 
> > You can merge this Pull Request by running
> > 
> >   git pull https://github.com/shaungrady/angular-ui mask
> > 
> > Or view, comment on, or merge it at:
> > 
> >   https://github.com/angular-ui/angular-ui/pull/479
> > Commit Summary
> > - Mask refactor, no more dependencies
> > - No more dependencies
> > 
> > File Changes
> > - _D_ modules/directives/mask/dependencies.jsonhttps://github.com/angular-ui/angular-ui/pull/479/files#diff-0(5)
> > - _M_ modules/directives/mask/mask.jshttps://github.com/angular-ui/angular-ui/pull/479/files#diff-1(341)
> > 
> > Patch Links:
> > - https://github.com/angular-ui/angular-ui/pull/479.patch
> > - https://github.com/angular-ui/angular-ui/pull/479.diff

shaungrady commented:
Hi Pete, 

Thanks! I agree, interpolation does make more sense. I will work on that today. I will also work on making some of the comments in the code more helpful. 

I'm not sure how to test UI-triggered events either. What do you think of this approach: 

On line 93, where the event listener is attached, adding a custom event…

```
iElement.bind('input propertychange keyup click mouseout maskdebug', eventHandler);
```

… and add a second argument to the eventHandler func for testing, have it do something like…

```
function eventHandler(e, eDebug) {
    var isDebug = (e.type == 'maskdebug') && angular.isObject(eDebug);
    e = isDebug ? eDebug : e;

    ...

    // At the end of the handler, hit the testing callback
    if (isDebug)
        e.callback(valMasked, caretPos);
```

Then, in the test, we can use jQuery's .triggerHandler() and pass our own hand-made event object to the handler.

```
function testCallback(valMasked, caretPos) {
    // test to see if caret moved leftwards passing over all mask-component chars
}
element.triggerHandler('maskdebug', { which:37, callback:testCallback });
```

What do you think?

petebacondarwin commented:
RE triggering - Maybe ...
I have changed the code on my branch to use interpolation already.  See the
unit tests.

On 11 March 2013 14:57, Shaun Grady notifications@github.com wrote:

> Hi Pete,
> 
> Thanks! I agree, interpolation does make more sense. I will work on that
> today. I will also work on making some of the comments in the code more
> helpful.
> 
> I'm not sure how to test UI-triggered events either. What do you think of
> this approach:
> 
> On line 93, where the event listener is attached, adding a custom event…
> 
> iElement.bind('input propertychange keyup click mouseout maskdebug', eventHandler);
> 
> … and add a second argument to the eventHandler func for testing, have it
> do something like…
> 
> function eventHandler(e, eDebug) {
>     var isDebug = (e.type == 'maskdebug') && angular.isObject(eDebug);
>     e = isDebug ? eDebug : e;
> 
> ```
> ...
> 
> // At the end of the handler, hit the testing callback
> if (isDebug)
>     e.callback(valMasked, caretPos);
> ```
> 
> Then, in the test, we can use jQuery's .triggerHandler() and pass our own
> hand-made event object to the handler.
> 
> function testCallback(valMasked, caretPos) {
>     // test to see if caret moved leftwards passing over all mask-component chars
> }
> element.triggerHandler('maskdebug', { which:37, callback:testCallback });
> 
> What do you think?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/angular-ui/angular-ui/pull/479#issuecomment-14717651
> .

shaungrady commented:
Ah I see, fantastic! Other than tests, is there anything else that needs to be done?

petebacondarwin commented:
Documentation - particular updating the gh pages index.html!
Also I was unsure about the input being cleared when the model is not
valid?  Shouldn't it look like it does when the masked input is first
rendered?
And also, I notice that you only update the model on blur.  This is not in
line with how other AngularJS input directives work.  Is there some reason
why it shouldn't be on every change?

On 11 March 2013 16:13, Shaun Grady notifications@github.com wrote:

> Ah I see, fantastic! Other than tests, is there anything else that needs
> to be done?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/angular-ui/angular-ui/pull/479#issuecomment-14722538
> .

shaungrady commented:
Input clearing on render:
I have mixed feelings about this. Even if the model is considered invalid by the mask directive on render, it seems wrong to have the a directive change the model without user interaction. The form would immediately be marked dirty and, if ngRequired, marked invalid. It feels more appropriate to leave the model untouched, even if it doesn't jive with the specified mask, until a user touches it. What are your thoughts on this?

The model is updated through controller.$parsers on every user input. It's not updated on blur, though, which is why I call $setViewValue to update the model to the cleared state (if mask was invalid when user blurred).

shaungrady commented:
```
This is not in line with how other AngularJS input directives work
```

Just to expand on my response on this, I believe it's the ngModel directive that's already pushing the input's value to the model; the mask directive needs only to add it's own formatter and parser. The $setViewValue in blurHandler makes sure the .val('') clearing on invalid mask goes through the parsing pipeline too. Is there a more appropriate way to handle this?

shaungrady committed:
Fix a bunch of bugs

Remember original placeholder and maxlength attrs, properly unbind
events, proper uninitialize func, proper clearing of value on blur with
zero-length raw val.

petebacondarwin commented:
This is all looking excellent now!  I think it is working as I would expect.

petebacondarwin commented:
Merged into master.

petebacondarwin commented:
Thanks Shaun for this cracking PR

shaungrady commented:
Thanks for the assistance Pete. There is another "hidden" feature of the mask where it can accept an array, allowing you to define your own RegEx patterns, as shown here: http://plnkr.co/edit/qKCgHmJqOg7iXczmup5i?p=preview

Though I'd like to think about it's implementation a bit more before really documenting it.

courtyenn commented:
Would anyone know if we can use this directive in the angular-ui-mask repository? I would like to send back the formatted value back to the server, and in the documentation, it isn't clear. Thanks!
